src/llmcompressor/modifiers/autoround/base.py

```diff
@@ -0,0 +1,335 @@
+            # Note: No need to register observers for auto-round
+            self._calibration_hooks |= self._initialize_hooks(module)
+            apply_calibration_status(module)
+
+        model.apply(enable_quantization)  # quantize at the same time as calibrate
+
+    def input_capture_hook(self, module, *args, **kwargs):
+        if module._tmp_name not in self._all_module_input:
+            self._all_module_input[module._tmp_name] = []
+        self._all_module_input[module._tmp_name].append((args, kwargs))
+
+    def on_start(self, state: State, event: Event, **kwargs):
+        self.started_ = True
+
+        # register quantization calibration hooks
+        # assume quantization has been initialized by this modifier or one before it
+        self.start_calibration(state.model)
+        for _, module in state.model.named_modules():
+            if self._is_decoding_layer(module):
+                # register input capture hook for decoding layers
+                self.register_hook(
+                    module, self.input_capture_hook, "forward_pre", with_kwargs=True
+                )
+
+    def on_event(self, state: State, event: Event, **kwargs):
+        if event.type_ == EventType.CALIBRATION_EPOCH_START:
+            if not self.started_:
+                self.on_start(state, None)
+
+        if event.type_ == EventType.SEQUENTIAL_EPOCH_END:
+            subgraph = kwargs.pop("subgraph", None)
+            self.apply_autoround(state, subgraph)
+            self.post_autoround_cleanup()
+
+        if event.type_ == EventType.CALIBRATION_EPOCH_END:
+            if not self.ended_:
+                self.on_end(state, None)
+
+    def apply_autoround(self, state, subgraph):
+        """
+        Applies AutoRound quantization tuning on the current decoding layer.
+
+        The tuning logic is as follows:
+        for iter in range(iters):
+            quant_output = forward(layer, cached_inputs)
+            loss = mse_loss(quant_output, original_output)
+            loss.backward()
+            optimizer.step()
+            if loss < best_loss:
+                best_params = update_params(layer)
+
+        For more details, please refer to the AutoRound repository:
+        https://github.com/intel/auto-round/
+        """
+        modules = list(subgraph.submodules(model=state.model))
+
+        decoding_layers = [m for m in modules if self._is_decoding_layer(m)]
+        if len(decoding_layers) == 0:
+            return
+        assert len(decoding_layers) == 1, (
+            "Only one decoding layer is expected in the subgraph, "
+            f"found {len(decoding_layers)}."
+        )
+        decoding_layer = decoding_layers[0]
+
+        logger.info("Applying AutoRound on layer {}", decoding_layer._tmp_name)
+
+        wrapped_model = _wrap_decoding_layer(decoding_layer)
+
+        with torch.enable_grad(), align_module_device(decoding_layer):
+            import auto_round
```

Comment:
can we move this to the top of the file? would be good to show imports to users at top scope